### PR TITLE
ES6 and Let's Encrpyt

### PR DESCRIPTION
**Pull request by Akxe**
Solves #11, #12 and #13 

- [x] Move to promises
- [x] Move to ES6 writing style
- [x] Variable hoisting (let, const)
- [x] Micro optimisations
- [ ] Moving from certBot to Let's Encrypt 
- [x] Make Router constructor configurable instead of file
- [ ] Export Router to world

Most of your code is commented for references, but parts that I knew how to rewrite is rewrited. All parts that are to be done are commented with TODO and what should be done. If you could help me out with the part, that is left, I would be very much pleased. I failed.

Now solves only #12, and partialy solves #11 and #13

### CODE REVIEW

**Comment by Akxe** [comment]
@Irrelon could you please review changes?

**Comment by Irrelon** [comment]
@Akxe Looks ok so far, cannot merge this if it doesn't work yet though. Other comments are inline in the code.

File: package.json [hunk 1]
```diff
@@ -1,7 +1,19 @@
 {
-  "name": "node-irrelon-router",
-  "version": "0.1.0",
+  "name": "node-akxe-router",
```
Comment (Irrelon) [inline package.json]: Should not change the name of the project unless you want to deploy this yourself to NPM.

File: package.json [hunk 2]
```diff
@@ -1,7 +1,19 @@
...
+  "version": "0.0.1",
```
Comment (Irrelon) [inline package.json]: Should not reverse the version numbers unless you are starting a complete fork that will never get merged back in.

File: package.json [hunk 3]
```diff
@@ -1,7 +1,19 @@
...
+    "greenlock": "^2.1.15",
+    "http-proxy": "^1.11.x",
+    "le-challenge-fs": "^2.0.8",
+    "le-challenge-sni": "^2.0.1",
+    "le-sni-auto": "^2.1.1",
+    "le-store-certbot": "^2.0.5",
+    "ssl-config": "^0.0.9"
+  },
+  "scripts": {
+    "start": "forever start -a --uid \"akxe-router\" router.js",
```
Comment (Irrelon) [inline package.json]: Name change here as well.

File: package.json [hunk 4]
```diff
@@ -1,7 +1,19 @@
...
+    "test": "node router.js",
+    "stop": "forever stop akxe-router"
```
Comment (Irrelon) [inline package.json]: Name change here as well.

File: router.js [hunk 5]
```diff
@@ -1,93 +1,501 @@
...
+const crypto = require('crypto')
+const tls = require('tls')
+const //express = require('express')
+const spawn = require('child_process').spawn
+const spawnSync = require('spawn-sync')
+const path = require('path')
+const fsAccess = require('fs-access')
+const mkdirp = require('mkdirp')
+const configFilePath = __dirname + '/config.json'*/
+
```
Comment (Irrelon) [inline router.js]: If this stuff is no longer needed, remove the code.

File: router.js [hunk 6]
```diff
@@ -1,93 +1,501 @@
...
+const greenLock = require('greenlock')
```
Comment (Irrelon) [inline router.js]: Semicolons have been pulled from the code completely. Semicolons remove ambiguity in code and should be in there. Seen too many errors in code caused by not adding them in the past from developers who are not as familiar with the code as you or I might be.

File: router.js [hunk 7]
```diff
@@ -1,93 +1,501 @@
...
+	static fromJSON(object){
+		return new Route(object)
+	}
+}
+
+const router = new Router({
+	email: 'akxe@example.com',
+	routesFile: './exampleConfig.json'
+})
+router.listen()
```
Comment (Irrelon) [inline router.js]: I'm thinking that given this is a complete re-write, some tests should be made to ensure this works. Including websocket tests as well as we are supposed to support them too.

File: package.json (hunk 1 above)
Comment (Akxe) [reply package.json]: Sorry my bad

File: router.js (hunk 6 above)
Comment (Akxe) [reply router.js]: Typically the IDE warns about semicolons missing.

File: router.js (hunk 7 above)
Comment (Akxe) [reply router.js]: Definitely agreed, however I am unable to resolve implementation issues with greenlock, thus it is not working with SSL. And second problem is that I don't know how to create tests, if you create one, I can copycat the others.

File: router.js (hunk 5 above)
Comment (Akxe) [reply router.js]: I removed a lot and probably missed that part my bad

**Pull request closed without merging**